rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+            of which has size B and the second of which has size C. shape should
+            consist of unique dimension names. For example having "B B" is invalid.
+        dtype: The dtype of the tensor. If None, the dtype is not checked during
+            validation. Also during Sampling the dtype is set the default dtype of
+            the backend.
+        shape_vals: An optional dictionary mapping some dimension names to their
+            values. For example, if shape is "B C" and shape_vals is {"C": 3}, then
+            the shape of the tensor is (B, 3). B is to be determined during
+            run-time but C is fixed to 3.
+
+    Examples:
+        >>> spec = TensorSpec("b h", h=128, dtype=tf.float32)
+        >>> spec.shape  # ('b', 128)
+        >>> spec.validate(torch.rand(32, 128, dtype=torch.float64))  # passes
+        >>> spec.validate(torch.rand(32, 64, dtype=torch.float64))   # raises ValueError
+        >>> spec.validate(torch.rand(32, 128, dtype=torch.float64))  # raises ValueError
+
+    Public Methods:
+        validate: Checks if the shape and dtype of the tensor matches the
+            specification.
+        sample: Samples a tensor with the specified value that is an
+            example of a tensor that matches the specification.
+
+    Abstract Methods:
+        get_shape: Returns the shape of the tensor depending on the backend.
+        get_dtype: Returns the dtype of the tensor depending on the backend.
+        _sample: Samples a tensor with the specified value that
+            has values of fill_value, shape of shape, and dtype of self.dtype.
+    """
+
+    def __init__(
+        self, shape: str, *, dtype: Optional[Any] = None, **shape_vals: Dict[str, int]
+    ) -> None:
+        self._expected_shape = self._parse_expected_shape(shape, shape_vals)
+        self._dtype = dtype
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """Returns a `tuple` specifying the tensor shape."""
+        return self._expected_shape
+
+    @property
+    def dtype(self) -> Any:
+        """Returns a dtype specifying the tensor dtype."""
+        return self._dtype
+
+    @DeveloperAPI
+    def validate(self, tensor: TensorType) -> None:
+        """Checks if the shape and dtype of the tensor matches the specification.
+
+        Args:
+            tensor: The tensor to be validated.
+
+        Raises:
+            ValueError: If the shape or dtype of the tensor does not match the
+        """
+        shape = self.get_shape(tensor)
+        if len(shape) != len(self._expected_shape):
+            raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        for expected_d, actual_d in zip(self._expected_shape, shape):
+            if isinstance(expected_d, int) and expected_d != actual_d:
+                raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        dtype = self.get_dtype(tensor)
```

Comment:
do we fail on dtype ... oh I guess this is important in the context of tensor types

Reply:
yes to kinda prevent float vs. double issues that are common when working with torch.